### PR TITLE
[torchgen] Add CI job to make sure torchgen works for Executorch op registration

### PR DESCRIPTION
**Pull request by larryliu0820**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89596

## Job

Test running on most CI jobs.

## Test binary

* `test_main.cpp`: entry for gtest
* `test_operator_registration.cpp`: test cases for gtest

## Helper sources

* `operator_registry.h/cpp`: simple operator registry for testing purpose.
* `Evalue.h`: a boxed data type that wraps ATen types, for testing purpose.
* `selected_operators.yaml`: operators Executorch care about so far, we should cover all of them.

## Templates

* `NativeFunctions.h`: for generating headers for native functions. (not compiled in the test, since we will be using `libtorch`)
* `RegisterCodegenUnboxedKernels.cpp`: for registering boxed operators.
* `Functions.h`: for declaring operator C++ APIs. Generated `Functions.h` merely wraps `ATen/Functions.h`.

## Build files

* `CMakeLists.txt`: generate code to register ops.
* `build.sh`: driver file, to be called by CI job.

### CODE REVIEW

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41506912).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41506912).

**Comment by ezyang** [comment]
If I have a stock build of PyTorch (non-CI), how do I run these tests? Is it really necessary to have an entirely different CI configuration? I feel like we should be roll it into a preexisting CI.

**Comment by larryliu0820** [comment]
> If I have a stock build of PyTorch (non-CI), how do I run these tests? Is it really necessary to have an entirely different CI configuration? I feel like we should be roll it into a preexisting CI.

That's the thing, I don't think PyTorch build should build Executorch. This should be a standalone build. In the future once we move all the Executorch types into open source, I think the CI job shouldn't even depend on PyTorch

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -1126,6 +1126,13 @@ install(FILES
   "${TORCH_SRC_DIR}/custom_class_detail.h"
   DESTINATION ${TORCH_INSTALL_INCLUDE_DIR}/torch)
 if(BUILD_TEST)
+  message("Building test. ${BUILD_EXECUTORCH}")
+  if(BUILD_EXECUTORCH)
+    add_subdirectory(
+            ${TORCH_ROOT}/test/edge
+            ${CMAKE_BINARY_DIR}/test_edge_op_registration
+    )
+  endif()
```
Comment (larryliu0820) [inline caffe2/CMakeLists.txt:1135]: @ezyang right now the PyTorch build includes Executorch build from here.

**Comment by ezyang** [comment]
I understand that, morally, it should be possible to build executorch without building pytorch, and I am fine with this existing as an option. However, as a developer, if executorch can be built as part of the standard pytorch build, it greatly reduces the barrier to entry for making changes to executorch, as I don't have to nuke my env and rebuild under some nonstandard configuration to repro an executorch problem.

**Comment by larryliu0820** [comment]
> I understand that, morally, it should be possible to build executorch without building pytorch, and I am fine with this existing as an option. However, as a developer, if executorch can be built as part of the standard pytorch build, it greatly reduces the barrier to entry for making changes to executorch, as I don't have to nuke my env and rebuild under some nonstandard configuration to repro an executorch problem.

Sure that sounds good to me, I think we can build & test Executorch in standard build, and we can always create new CI job later if necessary.

Should I set the `BUILD_EXECUTORCH` flag to true by default, so that it always build? How to run the test though?

**Comment by ezyang** [comment]
Yeah, true by default, then add executorch to one of the preexisting test calls in test.sh

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41506912).

**Comment by ezyang** [comment]
holler when ready for review

**Comment by larryliu0820** [comment]
@ezyang I fixed clang env jobs, I think it's ready for another round of review. Thanks!

File: test/edge/Evalue.h [hunk 2]
```diff
@@ -0,0 +1,472 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+namespace torch {
+namespace executor {
```
Comment (ezyang) [inline test/edge/Evalue.h:6]: So this is duplicated with fbcode-internal Executorch, is that right? Can we have instructions for (1) if you need to make changes to this file, what needs to happen on fbcode style and (2) under what circumstances can we delete this file and reunify them again?

Comment (ezyang) [reply test/edge/Evalue.h:6]: I guess I'm not going to review this for functionality

Comment (larryliu0820) [reply test/edge/Evalue.h:6]: Good point!

File: test/edge/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+add_custom_command(
+        COMMENT "Generating sources"
+        OUTPUT ${GEN_COMMAND_sources}
+        COMMAND ${GEN_COMMAND}
+        DEPENDS
+        ${TORCH_ROOT}/aten/src/ATen/native/native_functions.yaml
+        ${TORCH_ROOT}/aten/src/ATen/native/tags.yaml
+        ${TEST_ROOT}/templates/Functions.h
+        ${TEST_ROOT}/templates/NativeFunctions.h
+        ${TEST_ROOT}/templates/RegisterCodegenUnboxedKernels.cpp
```
Comment (ezyang) [inline test/edge/CMakeLists.txt:32]: you gotta add the torchgen py files as dependencies too, or it won't rebuild when python changesq

**Comment by ezyang** [comment]
Build system changes look fine. You had a LOT of files pulled in from internal. What's the timeline for deduping them.

**Comment by larryliu0820** [comment]
> Build system changes look fine. You had a LOT of files pulled in from internal. What's the timeline for deduping them.

I think the only duplicating files are `Evalue.h` and the templates. I think once we open source Executorch I can dedup them.

**Comment by ezyang** [comment]
what's the timeline for oss executorch :)

**Comment by larryliu0820** [comment]
> what's the timeline for oss executorch :)

I think Q2 2023 is the plan of record. cc @raziel

**Review by ezyang — APPROVED** [review]

**Comment by larryliu0820** [comment]
＠pytorchbot rebase

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42179581).

**Pull request closed without merging**